Fetch packages' licenses at build time

arnaldo2792 opened the pull request:
<!--
Tips:
- Please read CONTRIBUTING.md to understand our process and our requests for PRs.
- Please file an issue before creating a PR so we can discuss the change and confirm it's not already being worked on.
-->

**Issue number:**
N / A

**Description of changes:**

There are packages to which multiple licenses apply, depending on who distributes the software. For said packages, the licenses are fetched at build time when the `Licenses.toml` configuration file is present. The fetched licenses are stored in the `licenses` folder, used to retrieve the license of the special packages while they are built.

When the `Licenses.toml` file and the `licenses` folder are provided, they are copied to the build directory of the package. If either path is missing, empty configurations are provided. Packages that use the `Licenses.toml` file to retrieve their licensing information, will fail the build, since it is expected that the user will provide said configurations.

TODO:
- [x] Update `BUILDING.md` to document `Licenses.toml`

**Testing done:**
- With `Licenses.toml`:

```toml
[package]
spdx-id = "LicenseRef-<PACKAGE-LICENSE>"
licenses = [
  { path = "LicensesRef-<PACKAGE-LICENSE>.pdf", license-url = "<url-to-remote-license>" }
]
```
And with a spec file as follows:
```spec
%global spdx_id %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml spdx-id <package>)
%global license_file %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml path <package> -p ./licenses)

# ...

License: %{spdx_id}

# ...
%files <package>
%license %{license_file}
```

Verified the built rpm used the correct configurations

- Verified licenses files from `Licenses.toml` are retrieved and stored in `licenses`
- Verified the build doesn't fail for packages that don't use `Licenses.toml`
- Verified the build fails for packages that use `Licenses.toml`, and this is missing

**Terms of contribution:**

By submitting this pull request, I agree that this contribution is dual-licensed under the terms of both the Apache License, version 2.0, and the MIT license.

arnaldo2792 committed:
build: integrate bottlerocket-license-tool

Add support for the `Licenses.toml` file, used to provide licensing
information for some packages.
When the `Licenses.toml` file and the `licenses` folder are provided,
they are copied to the build directory of the package. If either path is
missing, empty configurations are provided. When this happens, packages
that use `Licenses.toml` will fail to build.

Signed-off-by: Arnaldo Garcia Rincon <agarrcia@amazon.com>

bcressey commented on BUILDING.md:
```diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
+#### Packeges licenses
+
+Most of the packages included in the base OS image are distributed using the licenses included in the upstream sources.
+However, there are some cases where multiple licenses could apply to a package, depending on who will ditribute the OS images.
```
```suggestion
However, in some rare cases there are multiple licenses that could apply to a package.
```

bcressey commented on BUILDING.md:
```diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
 
 (You can use variant and arch arguments together, too.)
 
+#### Packeges licenses
```
```suggestion
#### Package licenses
```

bcressey commented on BUILDING.md:
```diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
 
+#### Packeges licenses
+
+Most of the packages included in the base OS image are distributed using the licenses included in the upstream sources.
```
```suggestion
Most packages will include license files extracted from the upstream source archives.
```

bcressey commented on BUILDING.md:
```diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
+Most of the packages included in the base OS image are distributed using the licenses included in the upstream sources.
+However, there are some cases where multiple licenses could apply to a package, depending on who will ditribute the OS images.
+Bottlerocket's build system uses the `Licenses.toml` file in conjuction with the `licenses` directory to configure the licenses used for such special packages.
+You can use them to specify what licenses better suits your needs when you build and distribute your own Bottlerocket images.
```
```suggestion
```

webern replied:
Grammar nit

```suggestion
You can use them to specify what licenses better suit your needs when you build and distribute your own Bottlerocket images.
```

bcressey commented on BUILDING.md:
````diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
+In the previous example, it is expected that the file `the-license.txt` is present in `licenses`.
+You can retrieve the licenses from a remote endpoint, or the local filesystem if you specify the `license-url` field:
+
+```toml
+spdx-id = "SPDX-ID AND SPDX-ID-2" # Package with multiple licenses
````
```suggestion
```toml
[package]
spdx-id = "SPDX-ID AND SPDX-ID-2" # Package with multiple licenses
```

bcressey commented on BUILDING.md:
```diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
+licenses = [
+  # This file is copied from a file system, and will be saved as `path`
+  { license-url = "file:///path/to/spdx-id-license.txt", path = "spdx-id-license.txt" },
+  # This file is fetched from a https endpoint, and will be saved as `path`
```
```suggestion
  # This file is fetched from an https endpoint, and will be saved as `path`
```

bcressey commented on BUILDING.md:
````diff
@@ -92,6 +92,37 @@ cargo make -e BUILDSYS_ARCH=my-arch-here
+]
+```
+
+**Note:** the image build process will fail if the license for a package is missing, make sure you provide the configurations for the packages that require them.
````
This is probably unnecessary; we can talk about specific packages as they come up, and in theory the package logic could be done in a way that made the files optional.
```suggestion
```

webern replied:
nit

```suggestion
**Note:** the image build process will fail if the license for a package is missing.
Make sure you provide configurations for the packages that require them.
```

bcressey commented on Dockerfile:
```diff
@@ -68,6 +68,16 @@ WORKDIR /home/builder
 USER builder
 ENV PACKAGE=${PACKAGE} ARCH=${ARCH}
 COPY --chown=builder roles/${REPO}.root.json ./rpmbuild/BUILD/root.json
+# We attempt to copy `Licenses.toml` and `licenses` for the current build, otherwise
+# an empty file and a directory are created. Packages that use `bottlerocket-license-tool`
+# will fail the build, since we expect users to provide their own license
+RUN --mount=target=/host \
+  ( [ -f /host/Licenses.toml ] \
+  && cp /host/Licenses.toml ./rpmbuild/BUILD/ \
+  || touch ./rpmbuild/BUILD/Licenses.toml ) \
+  && ( [ -d /host/licenses ] \
+  && cp -r /host/licenses ./rpmbuild/BUILD/ \
+  || mkdir ./rpmbuild/BUILD/licenses )
```
Are we doing this purely to get a more descriptive error message when the package build fails?

arnaldo2792 replied:
TL;DR: yes, see comment above ^.

bcressey commented on Makefile.toml:
```diff
@@ -89,6 +89,11 @@ BUILDSYS_UPSTREAM_SOURCE_FALLBACK = "false"
 # if the license check fails.
 BUILDSYS_ALLOW_FAILED_LICENSE_CHECK = "false"
 
+# We require package licenses to be cached to build an image. If you're working on a
+# local change and don't have license information yet, you can run with `-e
+# BUILDSYS_ALLOW_FAILED_LICENSE_FETCH=true` to allow the build to continue even
+BUILDSYS_ALLOW_FAILED_LICENSE_FETCH = "false"
```
I don't think we'd want to continue if we try to retrieve a license for which we have a remote URL and fail - that'd just turn into a package build failure later, and the situation described in the comment here is easily fixed by amending `Licenses.toml` to drop the remote URL.

We're really more concerned about "unexpected" network fetches, where the data is supposed to be local but isn't for some reason. So I'd probably make this a `BUILDSYS_UPSTREAM_LICENSE_FETCH` boolean that gates whether we reach out to the different URLs - it's a source of nondeterminism for builds and should be opt-in.

bcressey commented on Makefile.toml:
```diff
@@ -605,6 +610,32 @@ docker run --rm \
+run_fetch_licenses="
+bottlerocket-license-tool -l /tmp/Licenses.toml fetch /tmp/licenses
+"
+set +e
+if [ ! -f "${BUILDSYS_ROOT_DIR}"/Licenses.toml ] || [ ! -d "${BUILDSYS_ROOT_DIR}"/licenses ]; then
+  echo "Skipping fetching licenses since either 'Licenses.toml' or 'licenses' doesn't exist"
+  exit 0
+fi
```
Elsewhere - roles, keys - we are generally willing to create the expected storage directory if it doesn't already exist. 
```suggestion
set +e
if [ ! -f "${BUILDSYS_ROOT_DIR}"/Licenses.toml ] ; then
  echo "Not fetching licenses since 'Licenses.toml' doesn't exist"
  exit 0
fi

mkdir -p "${BUILDSYS_ROOT_DIR}/licenses"
```

webern commented on Dockerfile:
```diff
@@ -68,6 +68,16 @@ WORKDIR /home/builder
+RUN --mount=target=/host \
+  ( [ -f /host/Licenses.toml ] \
+  && cp /host/Licenses.toml ./rpmbuild/BUILD/ \
+  || touch ./rpmbuild/BUILD/Licenses.toml ) \
```
Is this necessary given that you tested for the existence of /host/Licenses.toml before copying it to here? (Same goes for the directory /host/licenses)

```suggestion
  || touch ./rpmbuild/BUILD/Licenses.toml ) \
```

arnaldo2792 replied:
Yes, this is necessary to create an empty `Licenses.toml`.  In the spec file we do something similar to

```spec
%global spdx_id %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml spdx-id <package>)
%global license_file %(bottlerocket-license-tool -l $PWD/rpmbuild/BUILD/Licenses.toml path <package> -p ./licenses)
```

If neither `Licenses.toml` nor `licenses` exist, the build will fail with a `File not found error`. By providing an empty file, the user will get a different error: `Couldn't find configuration for package <blah>`

webern commented on Makefile.toml:
```diff
@@ -89,6 +89,11 @@ BUILDSYS_UPSTREAM_SOURCE_FALLBACK = "false"
 
+# We require package licenses to be cached to build an image. If you're working on a
+# local change and don't have license information yet, you can run with `-e
+# BUILDSYS_ALLOW_FAILED_LICENSE_FETCH=true` to allow the build to continue even
```
nit: sentence not completed

```suggestion
# BUILDSYS_ALLOW_FAILED_LICENSE_FETCH=true` to allow the build to continue even
```

arnaldo2792 commented:
Forced push includes:

- Skip license fetch if `BUILDSYS_UPSTREAM_LICENSE_FETCH` is `false`
- Always attempt to create `licenses` folder, and skip license fetch when `Licenses.toml` doesn't exist
- Addressed feedback in documentation

bcressey commented on Makefile.toml:
```diff
@@ -605,6 +609,40 @@ docker run --rm \
+  exit 0
+fi
+
+mkdir -p ${BUILDSYS_ROOT_DIR}/licenses
```
nit:
```suggestion
mkdir -p "${BUILDSYS_ROOT_DIR}/licenses"
```

bcressey commented on Makefile.toml:
```diff
@@ -605,6 +609,40 @@ docker run --rm \
+  exit 0
+fi
+
+if [ ! -f "${BUILDSYS_ROOT_DIR}"/Licenses.toml ] ; then
```
nit:
```suggestion
if [ ! -s "${BUILDSYS_ROOT_DIR}/Licenses.toml" ] ; then
```

bcressey commented on Dockerfile:
```diff
@@ -68,6 +68,16 @@ WORKDIR /home/builder
 USER builder
 ENV PACKAGE=${PACKAGE} ARCH=${ARCH}
 COPY --chown=builder roles/${REPO}.root.json ./rpmbuild/BUILD/root.json
+# We attempt to copy `Licenses.toml` and `licenses` for the current build, otherwise
+# an empty file and a directory are created. Packages that use `bottlerocket-license-tool`
+# will fail the build, since we expect users to provide their own license
```
```suggestion
# We attempt to copy `Licenses.toml` and `licenses` for the current build, otherwise
# an empty file and a directory are created so that `bottlerocket-license-tool` will
# fail with a more descriptive error message.
```

arnaldo2792 commented:
Forced push addresses feedback

arnaldo2792 commented:
Forced push fixes commit message

arnaldo2792 commented:
Forced push includes the `BUILDSYS_LICENSES_CONFIG_PATH` env variable, used to copy the `Licenses.toml` to the root sources' directory. 


```fish
fetch-licenses on  fetch-licenses [$] ❯ rm Licenses.toml

fetch-licenses on  fetch-licenses [$] ❯ cargo make fetch-licenses -e BUILDSYS_UPSTREAM_LICENSE_FETCH=true -e BUILDSYS_LICENSES_CONFIG_PATH=$HOME/Licenses.toml
[cargo-make] INFO - cargo make 0.32.17
[cargo-make] INFO - Build File: Makefile.toml
[cargo-make] INFO - Task: fetch-licenses
[cargo-make] INFO - Profile: development
[cargo-make] INFO - Running Task: setup
[cargo-make] INFO - Running Task: setup-build
[cargo-make] INFO - Running Task: fetch-sdk
[cargo-make] INFO - Running Task: fetch-toolchain
[cargo-make] INFO - Running Task: fetch-sources
[cargo-make] INFO - Running Task: fetch-vendored
[cargo-make] INFO - Running Task: fetch-licenses
[cargo-make] INFO - Build Done in 3.50 seconds.

fetch-licenses on  fetch-licenses [$] ❯ test -f Licenses.toml; and echo "Exists"
Exists

```
